File: LC/python/Rotate List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    # @param head, a ListNode
    # @param k, an integer
    # @return a ListNode
    def rotateRight(self, head, k):
        if head == None:
            return None
        if k == 0:
            return head
        
        dummy = ListNode(0)
        dummy.next = head
        tmp = dummy
        cnt = 0
        while tmp.next:
            tmp=tmp.next
            cnt += 1
        
        tmp.next = dummy.next
        
        step = cnt - (k%cnt)
        for i in range(step): #shift left step = k%cnt
            tmp=tmp.next
        head = tmp.next
        tmp.next = None
        
        return head

File: LC/java/Rotate List.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode rotateRight(ListNode head, int n) {
        if (head == null || head.next == null){
            return head;
        }
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode p = dummy;
        int cnt = 0, step;
        
        while(p.next!=null){
            p = p.next;
            cnt += 1;
        }
        
        p.next = dummy.next;
        
        step = cnt - (n % cnt);
        while(step > 0){
            p = p.next;
            step -= 1;
        }
        
        head = p.next;
        p.next = null;
        
        return head;
    }
}